tests/gui/TestGui.cpp

```diff
@@ -1499,7 +1499,83 @@ void TestGui::testDatabaseSettings()
     triggerAction("actionDatabaseSettings");
     QTRY_COMPARE(autosaveDelayCheckBox->isChecked(), true);
     QTRY_COMPARE(autosaveDelaySpinBox->value(), autosaveDelayTestValue);
+    QTest::mouseClick(dbSettingsButtonBox->button(QDialogButtonBox::Ok), Qt::LeftButton);
     QApplication::processEvents();
+
+    // test autosave delay
+
+    // 1 get current change time and init settings
+
+    QFileInfo fileInfo(m_dbFilePath);
+    QDateTime lastModified = fileInfo.lastModified();
+    config()->set(Config::AutoSaveAfterEveryChange, true);
+
+    // 2 create new entries
+
+    // 2.a) Click the new entry button and set the title
+    auto* entryNewAction = m_mainWindow->findChild<QAction*>("actionEntryNew");
+    QVERIFY(entryNewAction->isEnabled());
+
+    auto* toolBar = m_mainWindow->findChild<QToolBar*>("toolBar");
+    QVERIFY(toolBar);
+
+    QWidget* entryNewWidget = toolBar->widgetForAction(entryNewAction);
+    QVERIFY(entryNewWidget->isVisible());
+    QVERIFY(entryNewWidget->isEnabled());
+
+    QTest::mouseClick(entryNewWidget, Qt::LeftButton);
+    QCOMPARE(m_dbWidget->currentMode(), DatabaseWidget::Mode::EditMode);
+
+    auto* editEntryWidget = m_dbWidget->findChild<EditEntryWidget*>("editEntryWidget");
+    QVERIFY(editEntryWidget);
+    auto* titleEdit = editEntryWidget->findChild<QLineEdit*>("titleEdit");
+    QVERIFY(titleEdit);
+
+    QTest::keyClicks(titleEdit, "Test autosaveDelay 1");
+
+    // 2.b) Save changes
+    editEntryWidget->setCurrentPage(0);
+    auto* editEntryWidgetButtonBox = editEntryWidget->findChild<QDialogButtonBox*>("buttonBox");
+    QTest::mouseClick(editEntryWidgetButtonBox->button(QDialogButtonBox::Ok), Qt::LeftButton);
+
+    // 2.c) Wait half the time, make sure file was not modified yet
+    Tools::wait(autosaveDelayTestValue * 60 * 1000 / 2);
```

Comment:
For GUI tests waiting 1 min is too much. What about to define 1 min delay and wait for 30 sec?

Reply:
I will lower it to 1 min, the test would check the first value for 30 seconds, then the second value for another 1 min and 10 sec.
This is the lowest for real use case. I understand this saves CI time.
I could also set a value lower then 1 min but this would not represent a real use case and might hurt accuracy of true negative test as the time it takes the CI to write to file could reach the scales of the delay...

Reply:
You should use the MockClock instead of actually waiting. We don't want _any_ delays during tests.

Reply:
while trying to remove all delays i still have to wait for the file write, i can not tell in advance what that real delay will be for writing the files yet i must wait for the file to change (with a timeout).
Would it be good to keep 10 seconds for each file write test (30 seconds) or is there a faster way to test for the modification time?

Reply:
Listen for the database saved signal

Reply:
MockClock seems to tricker ASSERT: "dateTime.timeSpec() == Qt::UTC", i can't figure out whats wrong, at a glance it looks like Clock sets UTC...

Reply:
It's used a lot of other places, replicate what they did.

Reply:
I removed real delays except timeouts while waiting for signals. I think the timer is not effected by the mockup clock. I was thinking about making a public function that gets the timer to fire and call it during testing as a way of speeding it up.
Here is an example of the way i tried testing, this is similar to other parts in the project using the mockclock
```cpp
    delayTestClock = new MockClock();
    MockClock::setup(delayTestClock);
    int saveDelayTimeoutMs = 20 * 1000; // save delay timeout
    QSignalSpy writeDbSignalSpy(m_db.data(), &Database::databaseFileChanged);
...
...
    // 3.b) Test modification time after expected
    delayTestClock->advanceSecond(autosaveDelayTestValue * 60 / 2);
    QApplication::processEvents();
    if (!writeDbSignalSpy.wait(saveDelayTimeoutMs)) {
        QFAIL("AutosaveDelay took too long to save");
    }
```
even making a large advancement with mockclock does not help.